contracts/token/ERC721_Enumerable_base.cairo

```diff
@@ -0,0 +1,247 @@
+    ERC721_burn
+)
+
+from contracts.ERC165 import ERC165_register_interface
+
+#
+# Storage
+#
+
+@storage_var
+func all_tokens_len() -> (res: Uint256):
+end
+
+@storage_var
+func all_tokens_list(index: Uint256) -> (token_id: Uint256):
+end
+
+@storage_var
+func all_tokens_index(token_id: Uint256) -> (index: Uint256):
+end
+
+@storage_var
+func owned_tokens(owner: felt, index: Uint256) -> (token_id: Uint256):
+end
+
+@storage_var
+func owned_tokens_index(token_id: Uint256) -> (index: Uint256):
+end
+
+#
+# Constructor
+#
+
+func ERC721_Enumerable_initializer{
+        syscall_ptr : felt*, 
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }():
+    # register IERC721_Enumerable
+    ERC165_register_interface('0x780e9d63')
+    return ()
+end
+
+#
+# Getters
+#
+
+func ERC721_Enumerable_totalSupply{
+        syscall_ptr: felt*, 
+        pedersen_ptr: HashBuiltin*, 
+        range_check_ptr
+    }() -> (totalSupply: Uint256):
+    let (totalSupply) = all_tokens_len.read()
+    return (totalSupply)
+end
+
+
+func ERC721_Enumerable_tokenByIndex{
+        syscall_ptr: felt*, 
+        pedersen_ptr: HashBuiltin*, 
+        range_check_ptr
+    }(index: Uint256) -> (token_id: Uint256):
+    alloc_locals
+    # Ensures index argument is less than total_supply 
+    let (len: Uint256) = ERC721_Enumerable_totalSupply()
+    let (is_lt) = uint256_lt(index, len)
+    assert is_lt = 1
+
+    let (token_id: Uint256) = all_tokens_list.read(index)
+    return (token_id)
+end
+
+func ERC721_Enumerable_tokenOfOwnerByIndex{
+        syscall_ptr: felt*, 
+        pedersen_ptr: HashBuiltin*, 
+        range_check_ptr
+    }(owner: felt, index: Uint256) -> (token_id: Uint256):
+    alloc_locals
+    # Ensures index argument is less than owner's balance 
+    let (len: Uint256) = ERC721_balanceOf(owner)
+    let (is_lt) = uint256_lt(index, len)
+    assert_not_zero(is_lt)
+
+    let (token_id: Uint256) = owned_tokens.read(owner, index)
+    return (token_id)
+end
+
+#
+# Externals
+#
+
+func ERC721_Enumerable_mint{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(to: felt, token_id: Uint256):
+    _add_token_to_all_tokens_enumeration(token_id)
+    _add_token_to_owner_enumeration(to, token_id)
+    ERC721_mint(to, token_id)
+    return ()
+end
+
+func ERC721_Enumerable_burn{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(token_id: Uint256):
+    let (_from) = ERC721_ownerOf(token_id)
+    _remove_token_from_owner_enumeration(_from, token_id)
+    _remove_token_from_all_tokens_enumeration(token_id)
+    ERC721_burn(token_id)
+    return ()
+end
+
+func ERC721_Enumerable_transferFrom{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(_from: felt, to: felt, token_id: Uint256):
+    _remove_token_from_owner_enumeration(_from, token_id)
+    _add_token_to_owner_enumeration(to, token_id)
```

Comment:
if `to` is zero this will register the token under the zero address

Reply:
`ERC721_transferFrom` (the next line, I believe) already does a zero address check. Devs _should_ use the enumerable contract _and_ ERC721_base, but do you think we should include it here just in case?

Reply:
no, that's fine